extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -394,3 +425,458 @@ export class VisibilityHelper {
 function timeBase(time, baseTime) {
   return time >= baseTime ? time - baseTime : 0;
 }
+
+
+/**
+ * A base class for `VisibilityRootForDoc` and `VisibilityRootForEmbed`. The
+ * instance of this class corresponds 1:1 to `AnalyticsRoot`. It represents a
+ * collection of all visibility triggers declared within the `AnalyticsRoot`.
+ * @implements {../../../src/service.Disposable}
+ * @abstract
+ */
+export class VisibilityRoot {
```

Comment:
after reading the code, I tend to rename it to `VisibilityManager`, since it manages not only the RootModel, but also ElementModels.